From dffe8e79a3644cae3ce5b99904d068f0c9805426 Mon Sep 17 00:00:00 2001
From: Hans-Christian Egtvedt <hcegtvedt@atmel.com>
Date: Wed, 17 Oct 2007 14:14:19 +0200
Subject: [PATCH] Fix AVR32 optimized memcmp not returning negative values

This patch fixes the assembler code for the AVR32 memcmp optimized function. In
some cases when the compare between two bytes in the function overflowed, the
assembler code would return positive values for instead of correct negative
values.

Signed-off-by: Hans-Christian Egtvedt <hcegtvedt@atmel.com>
---
 libc/string/avr32/memcmp.S |   19 +++++++++++--------
 1 files changed, 11 insertions(+), 8 deletions(-)

diff --git a/libc/string/avr32/memcmp.S b/libc/string/avr32/memcmp.S
index 7359a64..ae6cc91 100644
--- a/libc/string/avr32/memcmp.S
+++ b/libc/string/avr32/memcmp.S
@@ -41,14 +41,17 @@ memcmp:
 	retal	0
 
 .Lfound_word:
-	psub.b	r9, r8, r9
-	bfextu	r8, r9, 24, 8
-	retne	r8
-	bfextu	r8, r9, 16, 8
-	retne	r8
-	bfextu	r8, r9, 8, 8
-	retne	r8
-	retal	r9
+	mov	len, 4
+
+2:	bfextu	r11, r9, 24, 8
+	bfextu	r12, r8, 24, 8
+	sub	r12, r11
+	retne	r12
+	lsl	r8, 8
+	lsl	r9, 8
+	sub	len, 1
+	brne	2b
+	retal	r12
 
 	.size	memcmp, . - memcmp
 
-- 
1.5.2.5

